quodlibet/library/file.py

```diff
@@ -413,3 +418,159 @@ def remove_roots(self, old_roots: Iterable[str]) -> Generator[None, None, None]:
             self.remove(removed)
         else:
             print_d(f"No tracks in {old_roots} to remove from {self._name}")
+
+
+class WatchedFileLibraryMixin(FileLibrary):
+    """A File Library that sets up monitors on directories at refresh
+    and handles changes sensibly"""
+
+    def __init__(self, name=None):
+        super().__init__(name)
+        self._monitors = {}
+        print_d(f"Initialised {self!r}")
+
+    def monitor_dir(self, path: Path) -> None:
+        """Monitors a single directory"""
+
+        # Only add one monitor per absolute path...
+        if path not in self._monitors:
+            f = Gio.File.new_for_path(str(path))
+            try:
+                monitor = f.monitor_directory(Gio.FileMonitorFlags.WATCH_MOVES, None)
+            except GLib.GError as e:
+                print_w(f"Couldn't watch {path} ({e})")
+                monitor = None
+            if not monitor:
+                return
+            handler_id = monitor.connect("changed", self.__file_changed)
+            # Don't destroy references - http://stackoverflow.com/q/4535227
+            self._monitors[path] = (monitor, handler_id)
+            print_d(f"Monitoring {path!r}")
+
+    def __file_changed(self, monitor, main_file: Gio.File,
+                       other_file: Optional[Gio.File], event_type) -> None:
+        try:
+            EventType = Gio.FileMonitorEvent
+            file_path = main_file.get_path()
+            if file_path is None:
+                return
+            song = self.get(file_path)
+            file_path = Path(normalize_path(file_path, True))
+            other_path = (Path(normalize_path(other_file.get_path(), True))
```

Comment:
Considering that there are more than these, I think the `Path(normalize_path(path, True))` part should become a helper function.

Reply:
Yep some clean up would be good (after this is done, it goes way further than just here)